### PR TITLE
feat(team): add team extra_settings + poe v2 controls

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

Sometimes we need to store information / settings / context for a team that is temporary and won't stay around forever. We've done this in the past with `plugins_opt_in` for example. The problem is that then we have a column that stays around crowding our table.

Feature flags can handle some of this but not all - particularly if a team wants to override a feature flag. It would be possible to do it with feature flags but very convoluted.

## Changes

Screenshot 2023-05-09 at 16 13 58

Here I'm suggesting we add a field `extra_settings` to the team model. A loose JSON object where we can add context that needs to persist at the team level but that shouldn't be a column.

The use case here is PoE v2 - allowing teams to opt into this or opt out if we've enabled it for them. I'll then add a notice to queries for teams that have PoE v2 enabled pointing them to this if they would rather see old results

## How did you test this code?

Manually updated the setting in the UI

### CODE REVIEW

File: posthog/models/team/team.py
Comment (yakkomajuri) [inline posthog/models/team/team.py:163]:
```suggestion
    # likely doesn't deserve a dedicated column. Can be used for things like settings and overrides
```

**Comment by neilkakkar** [comment]
This is _exactly_ the kind of thing that the new beta management feature supports.

> particularly if a team wants to override a feature flag

Tell me more? What does this mean?

---

The only gotcha though, is that you wouldn't be able to locally evaluate things, since you wouldn't know whether they've opted in or out per se. Which is a big enough deal perhaps that it makes sense to go this route.

Just wanted to make you aware of the option :)

File: posthog/migrations/0314_team_extra_settings.py
Comment (tiina303) [inline posthog/migrations/0314_team_extra_settings.py:6]: Based on the PG migration happening on the 12th lets not ship this before that.

**Review by tiina303 — COMMENTED** [review]
> particularly if a team wants to override a feature flag
Tell me more? What does this mean?

Just opting into/out of the beta features.

> The only gotcha though, is that you wouldn't be able to locally evaluate things, since you wouldn't know whether they've opted in or out per se. Which is a big enough deal perhaps that it makes sense to go this route.

@neilkakkar  can you elaborate more on the beta features plan as maybe we should change it, so it's possible for backend to use beta features too - having two separate systems doesn't really make sense

**Comment by neilkakkar** [comment]
Yep, opting in and out should work as expected. This gets priority over whatever you might do internally, like rolling out to 20% of people - if a user is within the rollout but opted out, they'll stay opted out.

> so it's possible for backend to use beta features too - having two separate systems doesn't really make sense

Agree on not having two separate systems! And it's possible for backend to use this, just **not with local evaluation** (not because there are system limitations, just because you wouldn't know on the backend whether the user has opted in or not)

**Comment by yakkomajuri** [comment]
@neilkakkar that's the reason we need this unfortunately. We need this to be locally evaluated - can't afford a /decide call for each insight calculation. 

Decided to go this route as we'd previously asked @EDsCODE about using beta management for this

File: posthog/migrations/0314_team_extra_settings.py
Comment (yakkomajuri) [reply posthog/migrations/0314_team_extra_settings.py:6]: why? how would this affect things?

File: posthog/migrations/0314_team_extra_settings.py
Comment (yakkomajuri) [inline posthog/migrations/0314_team_extra_settings.py:16]: curious as to how the migrations test is failing given this?

File: posthog/migrations/0314_team_extra_settings.py
Comment (tiina303) [reply posthog/migrations/0314_team_extra_settings.py:6]: the last PG migration failed, because there was a django migration that broke replication

**Comment by yakkomajuri** [comment]
will wait until we're stable on the pg front to merge

**Review by fuziontech — APPROVED** [review]
`blank=True, null=True`

`stamp=True`

**Review by mariusandra — APPROVED** [review]
Lgtm. And as long as we get the migration in, we're 👌

**Comment by yakkomajuri** [comment]
Damn, by the time my tests run each time we have a new migration, what are all these for??

**Comment by Twixes** [comment]
I once wanted to add loose column like this to the User model, but someone (I don't remember, Karl?) led me away from it on the basis of that easily becoming a random blob of data down the line. Which I agree with now – let's use this very selectively, since if we start throwing all new fields into `extra_settings` – which would be very easy – we're gonna have a poor man's MongoDB instead of Postgres.
And in terms of database load, a strictly typed column or three would likely be more efficient than a single JSON one, even if kept lingering, since BOOLEAN is so much more compact than the unstructured JSONB.

**Pull request merged**
